BDRProblems/deadock_detector.py

```diff
@@ -0,0 +1,128 @@
+from collections import defaultdict
+
+# For each strongly connected component, find all elementary
+# cycles. 
+def simple_cycles(G):
+    def _unblock(thisnode, blocked, B):
+        stack = set([thisnode])
+        while stack:
+            node = stack.pop()
+            if node in blocked:
+                blocked.remove(node)
+                stack.update(B[node])
+                B[node].clear()
+    G = {v: set(nbrs) for (v,nbrs) in G.items()} # make a copy of the graph
+    sccs = strongly_connected_components(G)
+    while sccs:
+        scc = sccs.pop()
+        startnode = scc.pop()
+        path=[startnode]
+        blocked = set()
+        closed = set()
+        blocked.add(startnode)
+        B = defaultdict(set)
+        stack = [ (startnode,list(G[startnode])) ]
+        while stack:
+            thisnode, nbrs = stack[-1]
+            if nbrs:
+                nextnode = nbrs.pop()
+                if nextnode == startnode:
+                    yield path[:]
+                    closed.update(path)
+                elif nextnode not in blocked:
+                    path.append(nextnode)
+                    stack.append( (nextnode,list(G[nextnode])) )
+                    closed.discard(nextnode)
+                    blocked.add(nextnode)
+                    continue
+            if not nbrs:
+                if thisnode in closed:
+                    _unblock(thisnode,blocked,B)
+                else:
+                    for nbr in G[thisnode]:
+                        if thisnode not in B[nbr]:
+                            B[nbr].add(thisnode)
+                stack.pop()
+                path.pop()
+        remove_node(G, startnode)
+        H = subgraph(G, set(scc))
+        sccs.extend(strongly_connected_components(H))
```

Comment:
The search for strongly connected components is to identify the query that can break the cycle. As explained in the previous reply, there's no need to do a dfs to identify cycles on the scc because scc already represents cycle/s.